bgpd/bgpd.c

```diff
@@ -4403,6 +4403,11 @@ struct peer *peer_lookup_dynamic_neighbor(struct bgp *bgp, union sockunion *su)
 		zlog_debug("%s Dynamic Neighbor added, group %s count %d",
 			   peer->host, group->name, dncount);
 
+	if (dncount == gbgp->dynamic_neighbors_limit) {
```

Comment:
Isn't this part already doing the job?

```c
	if (dncount >= gbgp->dynamic_neighbors_limit) {
		if (bgp_debug_neighbor_events(NULL))
			zlog_debug(
				"Dynamic Neighbor %pFX rejected - at limit %d",
				&prefix, gbgp->dynamic_neighbors_limit);
		return NULL;
	}
```

Reply:
Currently we only have debug logs for when dynamic nbrs are more than listen limit , connections will be rejected saying : 

2023/11/07 18:36:00 BGP: Dynamic Neighbor 151.1.1.2/32 matches group PG7 listen range 151.1.1.0/24
2023/11/07 18:36:00 BGP: Dynamic Neighbor 151.1.1.2 rejected - at limit 5

But we don't have any info logs to indicate the users  when exactly the listen limit was reached or
after which prefix it started to reject new connections.

Reply:
Then it's a warning, not info?

Reply:
> Then it's a warning, not info?

Considered it as info initially, but indeed, "warning" is a more accurate.
Have updated accordingly.

Thanks

Reply:
By the way, in what case will you hit this statement? Because the statement above just returns and this is just a dead code, correct?

```c
	if (dncount >= gbgp->dynamic_neighbors_limit) {
		if (bgp_debug_neighbor_events(NULL))
			zlog_debug(
				"Dynamic Neighbor %pFX rejected - at limit %d",
				&prefix, gbgp->dynamic_neighbors_limit);
		return NULL;
	}
```

Reply:
> By the way, in what case will you hit this statement? Because the statement above just returns and this is just a dead code, correct?
> 
> ```c
> 	if (dncount >= gbgp->dynamic_neighbors_limit) {
> 		if (bgp_debug_neighbor_events(NULL))
> 			zlog_debug(
> 				"Dynamic Neighbor %pFX rejected - at limit %d",
> 				&prefix, gbgp->dynamic_neighbors_limit);
> 		return NULL;
> 	}
> ```
> 
> So this zlog_warn() should be inside this IF block, IMO.

Thanks for bringing it up Donatas,
Just to clarify the execution flow here. 

The statement `Dynamic Neighbor %pFX rejected - at limit %d` is indeed hit when we attempt to add a new connection after reaching the limit set by **gbgp->dynamic_neighbors_limit**. 

At this point, the condition **(dncount >= gbgp->dynamic_neighbors_limit)** evaluates to true, and we return NULL, indicating rejection of the new connection.

However, the addition of zlog_warn() serves a slightly different purpose. 
It triggers when the dncount is incremented just after successfully adding the last connection. 
This increment occurs at line 4400 **gbgp->dynamic_neighbors_count = ++dncount;**. 

Subsequently, when we evaluate if dncount is equal to the specified limit, we add this warning to indicate that this was indeed the last peer successfully added before reaching the limit.

so IMO that piece of code isn't dead; rather, it serves as a marker for the last successful addition before reaching the limit, providing additional context to the user.

Reply:
Should it be `==` rather than `>=` in that case?